File: Linked List/21.Merge Two Sorted Lists/Solution.py
class Solution:
    # @param two ListNodes
    # @return a ListNode
    def mergeTwoLists(self, l1, l2):
        if not l1: return l2
        if not l2: return l1
        nhead = ListNode(0)
        p = nhead
        while l1 and l2:
            if l1.val <= l2.val:
                p.next = l1
                p = p.next
                l1 = l1.next
            else:
                p.next = l2
                p = p.next
                l2 = l2.next
        if not l1:
            p.next = l2
        if not l2:
            p.next = l1
        return nhead.next

File: Linked List/21.Merge Two Sorted Lists/Solution.java
/**
 *	合并两个有序链表，得到一个新的有序链表。
 *	类似于归并排序中的Merge操作，只不过归并排序中是对数组合并。
 *	题目地址：https://leetcode.com/problems/merge-two-sorted-lists/
 *  相似题目：Sort List，Merge k Sorted Lists，Merge Sorted Array
**/
public class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode pHead = new ListNode(0);  //注意伪头结点的使用
        ListNode current = pHead;
        ListNode p = l1, q = l2;
        
        while( p != null && q != null) {  //比较两个链表当前结点的大小，先插入小的
            if(p.val <= q.val) {  
                current.next = p;
                p = p.next;
            } else {
                current.next = q;
                q = q.next;
            }
            current = current.next;  
        }
        if( p != null ) current.next = p;  //连接余下部分
        if( q != null ) current.next = q;
        return pHead.next;
    }
}